pkg/cloud/externalconn/connection_kms.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package externalconn
+
+import (
+	"context"
+	"net/url"
+
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/errors"
+)
+
+type kmsConnectionContext struct {
+	kmsEnv cloud.KMSEnv
+}
+
+// ExternalStorageContext implements the ConnectionContext interface.
+func (k *kmsConnectionContext) ExternalStorageContext() cloud.ExternalStorageContext {
+	panic("kmsConnectionContext cannot be used for External Storage initialization")
```

Comment:
Maybe return an err explicitly? This may be to subtle for users, who would probably prefer to be told "This is a KMS external connection, You can't use this as a Storage."

Reply:
I'm happy adding an error return type, I've been told in the past to keep signatures simpler and rely on panics to code that if reached is clearly a programming error. I'm not worried about it being subtle, it'll crash the cockroach node with the panic if anyone tries to misuse this.

Reply:
Oh, this is programmer error? A user can't get this by trying to use a KMS external conn in a STORAGE external conn slot?

Reply:
Oh yikes you're right, really good catch. I'm convinced this should be an error, but additionally we should have a check before we attempt to `Dial`. We should make sure that if you're asking for a KMS then the details is actually of type KMS, and similarly for storage. Making the change and adding a test!

Reply:
I've addressed this in the `LoadConnection` library method. It now checks that the object is of the correct type.